### PR TITLE
feat(ruleset): allow require calls in Node.JS

### PR DESCRIPTION
**Pull request by P0lip**
**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

File: docs/guides/custom-functions.md [hunk 1]
```diff
@@ -183,7 +183,7 @@ export default (obj) => {
 };
 ```
 
-The following code won't work properly either:
+Require calls will work only in Node.js, therefore it's strongly discouraged to use them unless you don't plan to distribute your ruleset among other users:
```
Comment (nulltoken) [inline docs/guides/custom-functions.md]: >strongly discouraged to use them unless you don't plan

@P0lip This double negative statement is a tad hard to parse and be 100% sure we understand it right the first time.

How about something like this:

Require calls will work only in Node.js. They won't work in browser based context. Thus, it's highly recommended to not rely on require calls if you plan to distribute your ruleset.

File: docs/guides/custom-functions.md [hunk 2]
```diff
@@ -183,7 +183,7 @@ export default (obj) => {
...
 
  ```js
 const foo = require('./foo'); // require is not available (see note below)
```
Comment (nulltoken) [inline docs/guides/custom-functions.md]: @P0lip In the context of that PR, that comment "// require is not available (see note below)" becomes a little bit confusing

File: docs/guides/custom-functions.md (hunk 1 above)
Comment (philsturgeon) [inline docs/guides/custom-functions.md]:
```suggestion
Require calls will work only in Node.js, and will cause errors for anyone trying to use the ruleset in the browser. If your ruleset is definitely going to only be used in the context of NodeJS then using them is ok, but if you are distributing your rulesets to the public we recommend avoiding the use of `require()` to increase portability.
```

File: docs/guides/custom-functions.md (hunk 2 above)
Comment (philsturgeon) [reply docs/guides/custom-functions.md]:
```suggestion
const foo = require('./foo');
```

File: docs/guides/custom-functions.md (hunk 1 above)
Comment (philsturgeon) [reply docs/guides/custom-functions.md]: I made a suggestion, think I got it.

**Comment by philsturgeon** [comment]
Yay thank you @P0lip! Shall we get a 5.2 out so that ruleset exceptions can be used in Studio, the YAML trouble in Studio will be resolved, and @mkistler can have this feature to play with?

**Comment by P0lip** [comment]
@philsturgeon 
Yes, that's the plan! I'm thinking of Monday or Tuesday, as indeed we have ruleset exceptions in, as well as a couple of other fixes.

**Comment by RobPhippen** [comment]
@philsturgeon @P0lip This is very good news! This has been a deal-breaker for me adopting spectral (FYI I am working with @mkistler )

File: src/rulesets/evaluators.ts [hunk 3]
```diff
@@ -1,12 +1,79 @@
...
+    return stubRequire();
+  }
+
+  if (source === null) {
+    return require;
+  }
+
+  return proxyRequire(source);
+};
+
```
Comment (XVincentX) [inline src/rulesets/evaluators.ts]: @P0lip @philsturgeon 

This parts suggest that Spectral is going to load whatever code, assuming it's all in good intent. Practically, I could write a rule that claims to lint a part of a JSON document but on the other hand it's formatting my disk or sending my credentials.

I know that in the browser is not an issue, but for anybody deploying a Spectral on the server this could be a problem. I highly suggest to use [vm2](https://github.com/patriksimek/vm2) and to put a clear whitelist of the modules that can be used.

Disabling external requires would be great, disabling `fs` is probably be a good start. The best tradeoff would be to propagate these configuration values from the spectral configuration file so that the user will decide how much risk they want to put in an external rule.

Comment (XVincentX) [reply src/rulesets/evaluators.ts]: //cc @nulltoken

Comment (P0lip) [reply src/rulesets/evaluators.ts]: We've had a couple of discussions regarding sandboxing. I was also leaning towards running the code in an isolated environment, but eventually we decided not to.
The primary reason is that no linter really does it. It's kind of the same story as with dependencies. We all kind of trust them. So we do here.

Comment (P0lip) [reply src/rulesets/evaluators.ts]: https://stoplight-internal.slack.com/archives/CEUN2664A/p1581573511007600?thread_ts=1581573511.007600&cid=CEUN2664A

Comment (XVincentX) [reply src/rulesets/evaluators.ts]: Read the conversation. Yeah my point is — as long you run in from your computer I really do not mind. My concern is more if Spectral will be run on a server (Stoplight SaaS?) then it might become an issue.

Since we're not doing this at the moment I guess it's ok to sandbox this. Thanks for checking it out!

Comment (P0lip) [reply src/rulesets/evaluators.ts]: >  My concern is more if Spectral will be run on a server (Stoplight SaaS?) then it might become an issue.

The plan is to spawn Spectral instances in a worker, so we'll also benefit from that, as workers are already partly isolated.
In any case, we are not planning to run node in a privileged mode (as root on Unix OSses, admin rights on Windows), so this will also mitigate certain issues.
Either way, point taken. I believe it'll be really worthwhile to revisit it at some point.
Providing a sandbox, for instance as an extra feature, would be something parts of users would likely benefit from.
